nsd.c

```diff
@@ -303,28 +303,84 @@ figure_default_sockets(
 
 	*ifs = i + 1;
 	setup_socket(&(*udp)[i], NULL, udp_port, &ai[0]);
-	figure_socket_servers(&(*udp)[i], NULL, server_count);
+	figure_socket_servers(&(*udp)[i], NULL);
 	setup_socket(&(*tcp)[i], NULL, tcp_port, &ai[1]);
-	figure_socket_servers(&(*tcp)[i], NULL, server_count);
+	figure_socket_servers(&(*tcp)[i], NULL);
+}
+
+static int
+find_device(
+	struct nsd_socket *sock,
+	const struct ifaddrs *ifa)
+{
+	for(; ifa != NULL; ifa = ifa->ifa_next) {
+		if((ifa->ifa_addr == NULL) ||
+		   (ifa->ifa_addr->sa_family != sock->addr.ai_family) ||
+		   ((ifa->ifa_flags & IFF_UP) == 0 ||
+		    (ifa->ifa_flags & IFF_LOOPBACK) != 0 ||
+		    (ifa->ifa_flags & IFF_RUNNING) == 0))
+		{
+			continue;
+		}
+
+#ifdef INET6
+		if(ifa->ifa_addr->sa_family == AF_INET6) {
+			struct sockaddr_in6 *sa1, *sa2;
+			size_t sz = sizeof(struct in6_addr);
+			sa1 = (struct sockaddr_in6 *)ifa->ifa_addr;
+			sa2 = (struct sockaddr_in6 *)&sock->addr.ai_addr;
+			if(memcmp(&sa1->sin6_addr, &sa2->sin6_addr, sz) == 0) {
+				break;
+			}
+		} else
+#endif
+		if(ifa->ifa_addr->sa_family == AF_INET) {
+			struct sockaddr_in *sa1, *sa2;
+			sa1 = (struct sockaddr_in *)ifa->ifa_addr;
+			sa2 = (struct sockaddr_in *)&sock->addr.ai_addr;
+			if(sa1->sin_addr.s_addr == sa2->sin_addr.s_addr) {
+				break;
+			}
+		}
+	}
+
+	if(ifa != NULL) {
+		char *colon;
+		size_t len;
+
+		if((colon = strchr(ifa->ifa_name, ':')) != NULL)
+			len = (size_t)((uintmax_t)colon - (uintmax_t)ifa->ifa_name);
+		else
+			len  = strlen(ifa->ifa_name);
+		strlcpy(sock->device, ifa->ifa_name, len);
```

Comment:
You still need to check if the len fits in the size that is available in the `sock->device` buffer.

Reply:
True. Fixed.